dxp/demystifying_aui_script_and_liferay-frontend_component_tags.md

```diff
@@ -0,0 +1,114 @@
+# Demystifying `aui:script` and `liferay-frontend:component` tags
+
+On JSP context you may notice a huge number of `aui:script` and `liferay-frontend:component` tags.
+
+## `aui:script` tag
+
+### What it does:
+
+This taglib allow users to use and require another JS module/resources inside a JSP.
+
+This taglib have three possible use cases:
+
+1. Pure JS on the JSP
+2. Import a AUI/YUI module
+3. Import a ES6 module
+
+#### Pure JS on the JSP
+
+By just placing JavaScript code inside an `aui:script` tag, you can take advantage of optimizing scripts and place them to be made available in the final compressed JavaScript bundle, see [combo for more details](https: / /github.com/liferay/liferay-frontend-guidelines/blob/45bad0334cb467e45e68aee322a2a4bc0b6979d3/dxp/javaScript_minification.md#comboservlet). It is worth mentioning that the same rule applies for the following use cases.
+
+#### Import an AUI/YUI module using `use` property
+
+When using the `use` property, the AUI / YUI module listed in the` use` list will be imported to the script data. If you want to know a little more about ScriptData, see this [link on how resources are injected into DXP](https://github.com/liferay/liferay-frontend-guidelines/blob/ea185686db8d562ac11cc2b95d2e9e7d7a0c6547/dxp/resource_injection.md#scriptdata).
+
+You can import several YUI / AUI modules within the same `use` property, as in this example:
+
+```jsp
+<aui:script use="liferay-upload">
+...
+        var liferayUpload = new Liferay.Upload({
+...
+</aui:script>
+```
+
+Note in the AlloyUI components, coming from the alloy-ui repository, classes are made available from the global variable `A`, as in this [code snippet the Menu component is defined](https://github.com/liferay/alloy-ui/blob/master/src/aui-menu/js/aui-menu.js#L39]).
+
+However, components made available from the `frontend-js-aui-web` module will be under the global Liferay variable. Like this [example here, `Liferay.Upload`](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/upload.js#L1541), used in the code example above.
+
+#### Import an ES6 module using `require` property
+
+Allows users to require a ES6 module as the following example:
+
+Considering that we already have `metal-clipboard` and `metal-dom` dependencies listed [on our module package.json](https://help.liferay.com/hc/en-us/articles/360018159771-The-Structure-of-OSGi-Bundles-Containing-npm-Packages-), We have:
+
+```jsp
+<aui:script require="metal-clipboard/src/Clipboard">
+    new metalClipboardSrcClipboard.default();
+</aui:script>
+```
+
+Also, we can set an alias with the `as` word after the provided path. It will be looking like:
+
+```jsp
+<aui:script require="metal-dom/all/dom.js as dom">
+
+	function myFunction(element) {
+		dom.addClasses(element, 'hide');
+	}
+
+	Liferay.once('something', myFunction);
```

Comment:
Delete all of this and replace it with:

```js
// ... code that uses the `dom` binding (eg. `dom.addClasses`) ...
```

The idea is to keep extraneous and distracting details out of the example (ie. what does this function have to do with it? What does `Liferay.once`?

I'm also a bit skeptical of the use of `metal-dom` for examples like this at all. I don't think we want to encourage the use of it when there are already built-in DOM APIs that do that kind of task (adding a `'hide'` class) without a dependency. And given that IE 11's days are numbered, there is even less reason to clutter our documentation with examples of legacy patterns that we don't want to encourage.

Reply:
Yes, good insight w/ metal-dom example; I could use other example. Also, thanks for remembering for not providing very distracting examples. I'll update it!

Reply:
I just created this example:

```jsp
<aui:script require="my-custom-lib/src/path/main as myLib">
	// ... code that use some function from myLib like `myLib.someFunction();`
</aui:script>
```

with this output example:

```js
Liferay.Loader.require('my-custom-lib/src/path/main', function(myCustomLibSrcPathMain) {
	try {
		(function() {
			var myLib = myCustomLibSrcPathMain;
			var $ = AUI.$;var _ = AUI._;
			// ... code that use some function from myLib like `myLib.someFunction();`
		})();
	} catch (err) {
		console.error(err);
	}
}
```

What do you think?